### ISSUE TITLE
Save game from Agile wont load in original AGI interpreter

### ISSUE THREAD

**Issue by vafada** [lanceewing/agile#18]
save a game while playing Agile.

try loading that save game in Nagi. Nagi will show an error 

https://github.com/sonneveld/nagi

**Comment by lanceewing** [comment]
Interesting. I haven't tried that yet myself. Does it work with the original AGI interpreter? I haven't done a lot of testing backwards and forwards between AGILE and other interpreters with regards to the saved games. 

By the way, thanks for contributing to the project. I really appreciate it.

**Comment by vafada** [comment]
Save game from AGILE wont load on Steam's Kings Question 2016 collection. 

It crashes the game.

**Comment by vafada** [comment]
Save  Game from Steam's Kings Question 2016 collection. wont load on AGILE

**Comment by vafada** [comment]
@lanceewing I've been playing around this one and couldn't figure this one out. But here's what I found:

I created a save game using both AGILE and KQ2 on Steam (I assume original interpreter)

the save game is just loading KQ2, do nothing and save the game immediately

The crash was on this line:

https://github.com/lanceewing/agile/blob/2015c196ae387fdef0c2b6161da1610598d1ff8a/AGILE/Commands.cs#L348-L351

because it was loading bad data (View with loop 121!)

So I added debug statements on this line:

https://github.com/lanceewing/agile/blob/2015c196ae387fdef0c2b6161da1610598d1ff8a/AGILE/SavedGames.cs#L1052-L1055

When loading the save game from Steam.. this is the output:

```
add.to.pic0 = 3 //view num
add.to.pic1 = 121
add.to.pic2= 0
add.to.pic3= 62
add.to.pic4= 30
add.to.pic5 = 121
**************
add.to.pic0 = 0
add.to.pic1 = 67
add.to.pic2= 20
add.to.pic3= 121
add.to.pic4= 0
add.to.pic5 = 68
**************
```

And based on the doc: http://agiwiki.sierrahelp.com/index.php?title=AGI_Specifications:_Chapter_10_-_Savegame_Files#10.2_Saved_Game_Format

its trying to add two views to the picture

view 3 which is red riding hood (which doesn't even appear on room 1), with loop 121...
view 0 (ego) with loop 67

obviously these numbers are wrong so AGILE crashes with this stack trace:

```
Exception thrown: 'System.ArgumentOutOfRangeException' in mscorlib.dll
ex: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
ex AgileForm: System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
   at System.Collections.ArrayList.get_Item(Int32 index)
   at AGILE.AnimatedObject.get_Loop() in C:\Users\Mark\Downloads\agile\AGILE\AnimatedObject.cs:line 75
   at AGILE.AnimatedObject.get_Cel() in C:\Users\Mark\Downloads\agile\AGILE\AnimatedObject.cs:line 91
   at AGILE.AnimatedObject.AddToPicture(Byte viewNum, Byte loopNum, Byte celNum, Byte x, Byte y, Byte priority, Byte controlBoxColour, Int32[] pixels) in C:\Users\Mark\Downloads\agile\AGILE\AnimatedObject.cs:line 1223
   at AGILE.Commands.ReplayScriptEvents() in C:\Users\Mark\Downloads\agile\AGILE\Commands.cs:line 358
   at AGILE.Commands.ExecuteAction(Action action, Byte& newRoom, Boolean& exit) in C:\Users\Mark\Downloads\agile\AGILE\Commands.cs:line 1656
   at AGILE.Commands.ExecuteLogic(Int32 logicNum) in C:\Users\Mark\Downloads\agile\AGILE\Commands.cs:line 2128
   at AGILE.Interpreter.Tick() in C:\Users\Mark\Downloads\agile\AGILE\Interpreter.cs:line 180
   at AGILE.AgileForm.InterpreterTick(Object sender, EventArgs e) in C:\Users\Mark\Downloads\agile\AGILE\AgileForm.cs:line 540
```

Other observations 

- when restoring the AGILE created save game, there is no `add.to.pic` script buffer events. Meaning this block never got executed:
https://github.com/lanceewing/agile/blob/2015c196ae387fdef0c2b6161da1610598d1ff8a/AGILE/SavedGames.cs#L1045-L1059

- I can load KQ1 save game created from Steam to AGILE... and same as above.. there were no `add.to.pic` script buffer events in the save file

I've attached the save game files i used to test:

[KQ2.zip](https://github.com/lanceewing/agile/files/9894770/KQ2.zip)

sg1 = agile
sg5 = steam

Hope this helps.

**Comment by lanceewing** [comment]
Okay, thanks for the info. I'll have to read through that in detail, and refresh my memory of the saved game code and format.

I haven't looked yet, but I recall that there are certain parts of the saved game format that change in length depending on the AGI interpreter version. I can't remember if AGILE only supports certain AGI interpreter versions with regards to the saved game format, or if I ended up adding logic to support older interpreter versions. I suspect that I haven't made it work with other versions, so one question would be whether the KQ2 from Steam is using an older interpreter version. My current install (probably from GOG) uses interpreter version 2.411. I will need to check if the AGILE saved game restoring code supports that version.

I will start looking at this issue later today, starting with reproducing and debugging. Thanks for the KQ2.zip.

**Comment by vafada** [comment]
Opening both save games files in a hex editor shows `Version 2.1` on offset 716

**Comment by lanceewing** [comment]
What AGI version does the Steam version show when you enter the debug mode using ALT-D?

The version 2.1 shown in the first window that pops up is the KQ2 game version, then in the second window, it shows the "Adventure Game Interpreter Version". That second string is hard coded into the AGI interpreter executable rather than being in the game data.

**Comment by vafada** [comment]
![Capture](https://user-images.githubusercontent.com/285941/198876872-f300119d-63a6-45f3-a032-d5b099de22d4.PNG)

**Comment by lanceewing** [comment]
Cool, that's the same as my one.

I have worked out the reason it is currently failing. It is because, when loading the script buffer event section in, it isn't using the "current number of script events" field from earlier in the saved game, and so it attempts to load and process all the data stored in that section. The problem is that that section contains garbage still left over from the previous room, e.g. loading pictures 0x62 and 0x68, which are both intro/credit sequence pictures. There must be some add to pic events in there as well, in fact, yeah, I can see them in there.

I just need to change it to make use of this field value:

            // [323] 354 - 355(2 bytes) Current number of script event entries.
            int scriptEntryCount = (savedGameData[354] + (savedGameData[355] << 8));

Fixed. Was quite a quick change. I've just pushed it to master.

**Comment by lanceewing** [comment]
I haven't tried the original NAGI issue though. This fix above is for the Steam -> AGILE issue. 

The AGILE -> NAGI and AGILE -> Steam issues would still be broken.

**Comment by vafada** [comment]
Tested and it works! awesome... I wouldn't worry about NAGI...  I don't know how complaint NAGI is with the original interpreter

Let me test loading a save game created by AGILE to Steam kq2 👿

**Comment by vafada** [comment]
closing this issue as fixed

**Comment by vafada** [comment]
Actually let me reopen.. the save game from AGILE wont load in Steam.....

but the steam save game loads in AGILE...

**Comment by vafada** [comment]
@lanceewing Steam KQ1 can open AGILE created save game!

and vice versa, AGILE can open Steam KQ1 created save game

![Capture](https://user-images.githubusercontent.com/285941/198906696-d7236336-2413-4560-a7ce-2f558a9e435d.PNG)

**Comment by vafada** [comment]
@lanceewing I think I figure out the reason why AGILE save game crashes the original interpreter.

this is all based on KQ2 save game.

I believe there is a bug in `byte[] objectData = state.Objects.Encode();`

https://github.com/lanceewing/agile/blob/210864b4d052a198f3835ddc52e84acfde2275f8/AGILE/SavedGames.cs#L638

Analyzing the hex dump between AGILE and  original interpreter, I noticed that when saving a KQ2 save game in AGILE the 

`byte[] objectData = state.Objects.Encode();`

results an array with size `699`

I ran the same data `KQ2` object file on my fork of `jagi` and the array has size `601`


![InkedCapture3](https://user-images.githubusercontent.com/285941/198926705-4b9fe28a-19f1-4166-bb6b-bb16c2df63fc.jpg)


Analyzing the hex dump for the save file  Original Interpreter, it says the object section has `598` length `598 + 3 = 601` (the first 3 bytes is removed from the save file). So the Original Interpreter has an array of length `601` too before generating the save file:

![InkedCapture2](https://user-images.githubusercontent.com/285941/198927091-2eb3e6ac-1445-4868-b09a-d0ecc1e2846f.jpg)

Analyzing the hex dump for AGILE save game, show that the object section has the wrong length. AGILE has `696` for the length of the object section `699 - 3 = 696`:

![InkedCapture](https://user-images.githubusercontent.com/285941/198927331-00291de6-55bd-403d-a610-f354208fff49.jpg)

You can also see from both screenshots that the data between the object length data (2 bytes) and the ASCII text `Sword.Trident.Bouquet of Flowers......` are different....

I copied and paste the parts of object section data from the Steam save game and pasted to the AGILE generated save game and I was able to load it using original interpreter though its broke the interpreter but better than crashing it

![Capture](https://user-images.githubusercontent.com/285941/198928037-300a206f-8d98-4117-8558-c8eacdd9a1c1.PNG)


so I believe the `state.Objects.Encode()` is generating byte array that is not compatible for the save game data format

Based on decompiled code by visual studio, it seems that the function is doing two for-loops, thus duplicating some data.... and the first for-loop includes the room number but the second for-loop does not. I don't have access to the source code so this is all a guess

**Comment by vafada** [comment]
@lanceewing 

Here's a quick hack I did to make the save game work. See the comments on the code that explains why

This replaces `byte[] objectData = state.Objects.Encode();` in `SavedGames.cs`

This is for kq2

```
                Objects objects = state.Objects;

                MemoryStream memoryStream = new MemoryStream();
                int count = objects.Count;
                int num = count * 3;
                memoryStream.WriteByte((byte)((uint)num & 0xFFu));
                memoryStream.WriteByte((byte)((uint)(num >> 8) & 0xFFu));
                memoryStream.WriteByte(state.Objects.NumOfAnimatedObjects);
                for (int i = 0; i < count; i++)
                {                
                    AGI.Resource.Object agiObject = objects[i];
                    // use fake data since we lost the offsets after parsing the OBJECT file
                    int fakeData = 255; // hardcoded offset. Offset 255 is the `?` item
                    memoryStream.WriteByte((byte)((uint)fakeData & 0xFFu));
                    memoryStream.WriteByte((byte)((uint)(fakeData >> 8) & 0xFFu));
                    memoryStream.WriteByte(agiObject.Room);
                }

                // start at 49..... 0-48 are just "?".. in the actual raw data "?" (index 49 is the first data after the offsets)
                for (int i = 49; i < objects.Count; i++)
                {
                    AGI.Resource.Object agiObject = objects[i];
                byte[] bytes = Encoding.ASCII.GetBytes(agiObject.Name);
                    foreach (byte value in bytes)
                    {
                        memoryStream.WriteByte(value);
                    }

                    memoryStream.WriteByte(0);
                }
            
                
                byte[] objectData = memoryStream.ToArray();
```

So the correct fix to just to store the `rawData` in `AGI.Resource.Objects` and just get it when its time to save the game.

We are losing the `Item Entries` data (offsets) on the `OBJECT` after parsing the file.

```
Item Entries

The first two bytes of each entry are an offset to the location of the inventory item's text value. Just like the header offset, all offsets are taken from the start of the item table (not the start of the file). The third byte of each entry is the location ([room] number) for the item.
```

Also the `state.Objects` doesn't really represent the raw data... `state.Objects` represent translated data based on  `Item Entries`  and `Item Text Data` mapping

 `state.Objects` index 0-49 is `?` but in raw data looks something like this (this map is not sorted, in the raw data its `?, Sword, Trident,....`

![Capture](https://user-images.githubusercontent.com/285941/199018212-32730ea8-5faf-442a-8298-5d1fbae51771.PNG)

**Comment by lanceewing** [comment]
> @lanceewing Steam KQ1 can open AGILE created save game!
> 
> and vice versa, AGILE can open Steam KQ1 created save game

I'm not surprised that games using version 2.917 of the original AGI interpreter are working, as the default behaviour of the saved game format is based on version 2.917, which made me wonder whether the bug was related to KQ2 being version 2.411 and the saved game format not adjusting to meet that, in fact I suspect that AGILE will save in a 2.917 saved game compatible format regardless of what version the game that was loaded was using. I think that AGILE may need to use the game's interpreter version to decide what to save in certain parts of the format.

The difference between a 2.411 and 2.917 saved game shouldn't be much though. The "pushed position within the script event list" is one difference though. 

            // [1505] 1536(2 bytes) Stores a pushed position within the script event list
            // Note: Depends on interpreter version. 2.4xx and below didn't have push.script/pop.script, so they didn't have this saved game field.
            savedGameData[1536] = (byte)(state.ScriptBuffer.SavedScript);

At the moment, there isn't a condition around that, so there will be a field present that the 2.411 isn't expecting. Not sure if it would simply be ignored, or if it could cause an issue. I suspect it may just ignore it, but it is a difference.

**Comment by lanceewing** [comment]
> @lanceewing I think I figure out the reason why AGILE save game crashes the original interpreter.

> results an array with size `699`
>
> I ran the same data `KQ2` object file on my fork of `jagi` and the array has size `601`

> so I believe the `state.Objects.Encode()` is generating byte array that is not compatible for the save game data format

Yeah, I think you are correct, that this size difference is causing the problem. The original AGI interpreter will have allocated memory based on the size of the OBJECT file that it loaded on game startup, and then it is making a valid assumption that the data in the saved game file is the same size, and so simply loads it into the same memory location, since the AGI saved game OBJECT section is essentially a memory dump of the in memory OBJECT file (**Edit:** which the AGI interpreter modifies in memory whenever an item's room has changed, so will most likely have changed).

So loading the additional 98 bytes into memory would cause it to overwrite something else.

As an aside, the AGI interpreter is known for not checking such memory boundaries, in fact some people have taken advantage of that to add additional 8086 machine code into the interpreter, using the set.string command, which likewise doesn't have boundary checks. That is obviously non standard, but it hasn't stopped people from extending the AGI interpreter through such code injection hacks. It would only work on one platform though, e.g. an 8086 platform.

**Comment by lanceewing** [comment]
> Here's a quick hack I did to make the save game work. See the comments on the code that explains why

> So the correct fix to just to store the `rawData` in `AGI.Resource.Objects` and just get it when its time to save the game.
> 
> We are losing the `Item Entries` data (offsets) on the `OBJECT` after parsing the file.

> Also the `state.Objects` doesn't really represent the raw data... `state.Objects` represent translated data based on `Item Entries` and `Item Text Data` mapping
> 
> `state.Objects` index 0-49 is `?` but in raw data looks something like this (this map is not sorted, in the raw data its `?, Sword, Trident,....`

At the moment I haven't looked to see why the AGILibrary is generating OBJECT data that is substantially longer than the original in this case. The code is from the "Visual AGI" project, which was a partially complete AGI IDE, similar to WinAGI and AGI Studio. It only had the capability of editing pictures, views, the object file, and words. So the OBJECT code to encode has the goal of creating an actual OBJECT file on disk, as part of game creation. It wouldn't have had in mind the idea of retaining the data size and data placement of the original OBJECT file, in the scenario where no inventory items had been added, deleted, or modified by the Visual AGI IDE. Obviously if modifications have been made, then it needs to regenerate with modified offsets, etc.

I will have a go at implementing a mechanism in the AGI Library to retain the size and original structure, i.e. the original raw data. Obviously it isn't the original raw data though, as the inventory item room locations may have changed, so it will need to take that into consideration.

**Comment by lanceewing** [comment]
> At the moment I haven't looked to see why the AGI Library is generating OBJECT data that is substantially longer than the original in this case.

Oh, I see now. It is all those extra "?" names in there. So AGILE is outputting a separate "?" name for each of the inventory objects that have that name, whereas the original data file was sharing the same name text for all of those objects with that name. Maybe the original AGI OBJECT file compiler utility was coded like that, to reuse text if multiple objects had the same name.

I'm not sure why KQ2 has this long list of "?" objects. I assume it is some kind of placeholder section that they had in the game. It doesn't appear to be a standard things across games, so perhaps it was the specific developers working on KQ2 that arranged their object file like this. This placeholder section is 50 items. Maybe it was a way to make the inventory item numbers start from 50 for this game.

**Comment by vafada** [comment]
> At the moment I haven't looked to see why the AGILibrary is generating OBJECT data that is substantially longer than the original in this case. The code is from the "Visual AGI" project, which was a partially complete AGI IDE, similar to WinAGI and AGI Studio. It only had the capability of editing pictures, views, the object file, and words. So the OBJECT code to encode has the goal of creating an actual OBJECT file on disk, as part of game creation. It wouldn't have had in mind the idea of retaining the data size and data placement of the original OBJECT file, in the scenario where no inventory items had been added, deleted, or modified by the Visual AGI IDE. Obviously if modifications have been made, then it needs to regenerate with modified offsets, etc.
> 

**Edit**: nevermind.. did not saw your post before.. but yeah.. its those pesky extra 49 `?` chars

the extra data is the ` 0-48` index in the `state.Objects` which is the `?` character and a NULL char 

`49 * 2 = 98` = 49 entries multiply by 2 (`?\0`) bytes

which matches the math `696 - 598 =  98` = 696 array size generated by `Encode()` , 598 array size by original interpreter.

Let me try opening a PR that fixes the `Encode` issues

**Comment by vafada** [comment]
> AGI saved game OBJECT section is essentially a memory dump of the in memory OBJECT file.

This is not true.... when you pick up an object and save the game, the save game includes the new room number

in kq2... 

before picking up the Trident, using original AGI

offset decimal 2424 has `07 01 24`

after picking up the trident

offset decimal 2424 becomes `07 01 FF`

WinAGI also states:

`An unencrypted copy of the game's OBJECT file data including any modifications to room information made by get, put and drop commands`

**Comment by vafada** [comment]
> The difference between a 2.411 and 2.917 saved game shouldn't be much though. The "pushed position within the script event list" is one difference though.

should we open a ticket for this? this is quite easy to fix.. we know the interpreter version `this.Version`

I noticed that too in the kq2 AGILEsave file (using v 2.4)..... at some point all the data is moved 2 bytes compared to the save file from original

**Comment by lanceewing** [comment]
> I will have a go at implementing a mechanism in the AGI Library to retain the size and original structure, i.e. the original raw data. Obviously it isn't the original raw data though, as the inventory item room locations may have changed, so it will need to take that into consideration.

Okay, that's done now. The Encode method in the AGILibrary now produces the raw byte data that exactly matches the original input date. It looks like I ended up with something quite similar to your code. I think the only difference is that I'm doing the storing of the name to name offset mapping within the same loop that outputs the name offsets and room number, rather than in a separate loop.

**Comment by lanceewing** [comment]
I also pushed a fixed to SavedGames.cs for the NumOfAnimatedObjects value. The restore method was using the number of animated object slots in the saved game file to set NumOfAnimatedObjects, but actually this should have had 1 subtracted from it, as the saved game file contains an additional animated object slot that is used for the add.to.pic. So when repeating save, restore, save, restore, etc., the value in the saved game file kept incrementing by 1. Fixed now though.

**Comment by lanceewing** [comment]
Regarding encryption of the OBJECT data section in the saved game file, WinAGI claims that this is encrypted for V3 games. AGILE isn't doing this at the moment, so that implies that V3 saved games won't work when loaded into the original V3 AGI interpreter. For example, for KQ4AGI, Gold Rush, the Manhunter games, and the V3 AGI version of Black Cauldron.

**Comment by lanceewing** [comment]
> > The difference between a 2.411 and 2.917 saved game shouldn't be much though. The "pushed position within the script event list" is one difference though.
> 
> should we open a ticket for this? this is quite easy to fix.. we know the interpreter version `this.Version`
> 
> I noticed that too in the kq2 AGILE save file (using v 2.4)..... at some point all the data is moved 2 bytes compared to the save file from original

Yeah, I'll open a ticket for that one now.

I'll also add a ticket for the V3 OBJECT section encryption.

**Comment by vafada** [comment]
Confirmed fixed! thanks

This fix spawn this new issue: #34
